### PR TITLE
Add support for compact and masked watersheds

### LINKED ISSUE #34: Speed-up watershed segmentation using information from a binary image
Following the [example from the documentation](https://juliaimages.org/latest/imagesegmentation/#Watershed-1), currently segmentation is performed on the whole image, and later it's filtered with the binary mask. Though, using mask during the segmentation would allow to significantly reduce computation time, avoiding computation for background pixels.

How about adding new optional parameter `img_bin::Union{Array{Bool,2}, BitArray{2}, Nothing}=nothing` to the function? So the interface will be changed from

```julia
watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N})
```
to
```julia
watershed(img::AbstractArray{T, N}, markers::AbstractArray{S,N}, img_bin::Union{AbstractArray{Bool,2}, BitArray{2}, Nothing}=nothing)
```

On my example, new function works 3.5 times faster. If you think it's fine, I will create a pull request.

### PR DESCRIPTION
**Pull request by tlnagy**
This PR adds support for the compact watershed algorithm (https://www.tu-chemnitz.de/etit/proaut/publications/cws_pSLIC_ICPR.pdf). This algorithm is nice because it results in more uniform watershed basins as you increase the `compactness` parameter. 

I still need to do the following before I think it should be merged:

- [x] Add tests
- [x] More docs (and reference!)
- [x] Benchmarking (thanks to https://github.com/scikit-image/scikit-image/issues/2636 for speed enhancements)

Example segmentation following the coin example from the docs:

| Simple Watershed  | Compact watershed  |
|---|---|
| simple | compact |

closes #34

### CODE REVIEW

**Comment by tlnagy** [comment]
~~Looks like my PR introduces quite a few allocations which are causing some slowdowns. Not sure where they are creeping in just yet.~~ Looks like I had some parameterization problems, the updated PR has the same performance as master

### Simple watershed

Master

```julia
julia> @benchmark segments = watershed(dist, markers)
BenchmarkTools.Trial: 
  memory estimate:  3.24 MiB
  allocs estimate:  119
  --------------
  minimum time:     41.722 ms (0.00% GC)
  median time:      42.512 ms (0.00% GC)
  mean time:        43.070 ms (0.92% GC)
  maximum time:     48.598 ms (5.49% GC)
  --------------
  samples:          117
  evals/sample:     1
```

This PR

```julia
julia> @benchmark segments = watershed(dist, markers)
BenchmarkTools.Trial: 
  memory estimate:  3.37 MiB
  allocs estimate:  119
  --------------
  minimum time:     44.658 ms (0.00% GC)
  median time:      45.398 ms (0.00% GC)
  mean time:        45.813 ms (0.92% GC)
  maximum time:     48.927 ms (5.84% GC)
  --------------
  samples:          110
  evals/sample:     1
```

**Comment by timholy** [comment]
Sorry for the delay here. It will take me a little while to read the paper so that I can understand this. Once you get eager to have this merged strip the WIP and ping me.

**Comment by tlnagy** [comment]
The TL;DR on the compact watershed algorithm is that, instead of assigning the final label as a pixel is pushed into the queue, it updates the label every time it encounters that position if the new label is more optimal than the previous label. More optimal in this case is a linear combination of the euclidean distance between the seed and the pixel and the typical brightness of the pixel. It doesn't lock in the final labelling until the pixels come off the queue, this lets it find better labelings later and replace initial labelings, which cannot happen with standard watershed.

I hope that helps explain what happens.

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]: Perhaps add a citation?

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]: Add a description of `mask`?

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]:
```suggestion
                   mask::AbstractArray{Bool, N}=fill(true, axes(img)),
```

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]: I'm actually unsure why we even add the type parameter at all. Is it bad to potentially compare `PixelKey`s with different `val` types?

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]:
```suggestion
        # when it comes off the queue since we could find a better suited watershed later.
```
(just making the tenses consistent, if this isn't the right change ignore it)

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]: this seems incomplete or confusing. Do we NOT set the grouping at push-time and also NOT calculate a weighted value...? Or is calculating the weighted value something we do INSTEAD?

**Review by timholy — DISMISSED** [review]
Happy to merge this whenever you want.

File: src/watershed.jl
Comment (tlnagy) [reply src/watershed.jl]: I decided that comparing different values is fine. We only need dimensional parameter so we don't accidentally compare pixels with different dimensions

File: src/watershed.jl
Comment (tlnagy) [reply src/watershed.jl]: Fixed.

**Comment by tlnagy** [comment]
I rebased, fixed the PR based on your suggestions, fixed a representation error bug with segmenting fixed point number images, and added tests. 

I think this is ready to merge once the tests pass.

**Comment by tlnagy** [comment]
@timholy Whenever you get a chance, could you merge this and tag a new release? Thanks!

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]: Check `compact` first since that will be quick

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]: Removing the type-parameter here seems likely to cause a performance hit: the priority queue needs to order its items, and the `isless` method accesses `val`. What was the motivation for removing it rather than just adding a second parameter `N`? Was it causing problems? If so can we address this in another way?

**Review by timholy — COMMENTED** [review]
Looks great! Any chance we could see fully up-to-date benchmarks, or has it not changed from what you posted?

I did suggest a couple of changes that seem likely to yield performance improvements.

File: src/watershed.jl
Comment (timholy) [reply src/watershed.jl]: In my comment I meant adding the type parameter to this method, not to the struct. Meaning it could be defined `isless(a::PixelKey, b::PixelKey) = ...` But if you need the `N` to match you can either reorder the parameters or use `PixelKey{S,N}` and `PixelKey{T,N}`.

**Comment by johnnychen94** [comment]
Reminder: retrigger the test after https://github.com/JuliaImages/ImageSegmentation.jl/pull/44 being merged. (Done)

**Comment by tlnagy** [comment]
I updated this PR by adding back the type paramerization in PixelKey. This fixes the performance regressions, it's basically back to the original speed. The compact algorithm now always uses floats to avoid the representation error when accumulating using fixed point numbers. @timholy and @johnnychen94, I believe this is ready to merge.

This PR:

```julia
julia> @benchmark segments = watershed(dist, markers)
BenchmarkTools.Trial: 
  memory estimate:  3.44 MiB
  allocs estimate:  121
  --------------
  minimum time:     35.417 ms (0.00% GC)
  median time:      39.968 ms (0.00% GC)
  mean time:        39.711 ms (0.11% GC)
  maximum time:     42.989 ms (0.00% GC)
  --------------
  samples:          126
  evals/sample:     1
```

Master:

```julia
julia> @benchmark segments = watershed(dist, markers)
BenchmarkTools.Trial: 
  memory estimate:  3.24 MiB
  allocs estimate:  119
  --------------
  minimum time:     33.161 ms (0.00% GC)
  median time:      36.747 ms (0.00% GC)
  mean time:        36.770 ms (0.11% GC)
  maximum time:     41.609 ms (0.00% GC)
  --------------
  samples:          136
  evals/sample:     1
```

**Comment by tlnagy** [comment]
Not sure why Travis isn't showing up under checks, but the build passed: https://travis-ci.org/github/JuliaImages/ImageSegmentation.jl/builds/672668131

**Comment by tlnagy** [comment]
Also this should fix #34

File: src/watershed.jl
Comment (johnnychen94) [inline src/watershed.jl]: Instead of hardcoded `Float64`, we can use `floattype(T)`, which is almost equivalent to `float(T)`(https://github.com/JuliaMath/FixedPointNumbers.jl/issues/127). If colorant wrapper is not wanted, then it could be `floattype(eltype(T))`.
Other usages of `Float64` can be changed accordingly.

File: test/watershed.jl
Comment (johnnychen94) [inline test/watershed.jl]: Does `watershed(Gray.(img), markers, compactness=10.0, mask=mask)` works normally, and how about `watershed(Gray.(ing), markers)`?

**Review by johnnychen94 — COMMENTED** [review]
Not familiar with the algorithm details, but it generally looks good to me.

I find it doesn't have example codes in the docstring, can you add one for it?

> Not sure why Travis isn't showing up under checks

It seems to have been broken for about a month in the dark time across all repos.

File: src/watershed.jl
Comment (johnnychen94) [inline src/watershed.jl]:
```suggestion
                   compactness::Real = 0.0) where {T<:Images.NumberLike, S<:Integer, N}
```

`Float64` could be problematic:

```julia
julia> function foo(; x::Float64)
       @show x
       end
foo (generic function with 1 method)

julia> foo(x=1.0)
x = 1.0
1.0

julia> foo(x=1.0f0)
ERROR: TypeError: in keyword argument x, expected Float64, got Float32
Stacktrace:
 [1] top-level scope at REPL[21]:1
```

File: src/watershed.jl
Comment (timholy) [reply src/watershed.jl]: I would also define this as a standalone variable
```
PK = PixelKey{floattype(eltype(T)),N}
pq = PriorityQueue{CartesianIndex{N}, PK}
```
and then use `PK(img[i], time_step, j)` below.

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]: Is the `img[1]` a bug? Should it be `img[i]`? See also the `PK` suggestion above.

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]: Should also make this `PK(new_value, time_step, curr_elem.source)`.

File: src/watershed.jl
Comment (timholy) [inline src/watershed.jl]: Likewise

**Review by timholy — COMMENTED** [review]
Sorry for the delay. I like @johnnychen94's comments, if we can address the last bit let's get this merged!

File: src/watershed.jl
Comment (tlnagy) [reply src/watershed.jl]: Good catch...

File: src/watershed.jl
Comment (tlnagy) [reply src/watershed.jl]: I loosened the type signature.

**Comment by tlnagy** [comment]
@johnnychen94 and @timholy I finally got around to making all the suggested changes. Hope this is finally ready for a merge!

**Review by johnnychen94 — APPROVED** [review]
LGTM, will merge in three days if there's no further comments

**Comment by tlnagy** [comment]
Bump @johnnychen94 and @timholy. Would be great to have this merged!

**Comment by johnnychen94** [comment]
Thanks for the work and patience @tlnagy !

**Comment by tlnagy** [comment]
Could you tag a new version when you get a chance? Thanks!

**Comment by timholy** [comment]
Thanks @tlnagy, and @johnnychen94 for merging!

**Pull request merged**
